### PR TITLE
Pep 593 py3

### PR DESCRIPTION
**Pull request by till-varoquaux**
This covers python 3.53 and above. I added a fair amount of tests; this diff is hard to read in this interface because it doesn't seem to offer a way to ignore whitespaces.

### CODE REVIEW

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 1]
```diff
@@ -1474,179 +1479,244 @@ def test_total(self):
...
-            self.assertEqual(
-                {Annotated[int, 4, 5], Annotated[int, 4, 5], Annotated[T, 4, 5]},
-                {Annotated[int, 4, 5], Annotated[T, 4, 5]}
-            )
-
-        def test_instantiate(self):
-            class C:
-                classvar = 4
+    def test_cannot_getattr_typevar(self):
+        with self.assertRaises(Exception):
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]: Should this be more specific? Maybe `AtttrubuteError`?

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 2]
```diff
@@ -1474,179 +1479,244 @@ def test_total(self):
...
+            classvar = 4
 
-            A = Annotated[C, "a decoration"]
-            a = A(5)
-            c = C(5)
-            self.assertEqual(a, c)
-            self.assertEqual(a.x, c.x)
-            self.assertEqual(A.classvar, C.classvar)
+        A = Annotated[C, "a decoration"]
+        self.assertEqual(A.classvar, 4)
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]: Could you also test assigning a class variable? I.e. set it on `A`, and then read on `C`.

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 3]
```diff
@@ -1474,179 +1479,244 @@ def test_total(self):
...
 
-            for t in samples:
-                x = Annotated[t, "a"]
 
-                for prot in range(pickle.HIGHEST_PROTOCOL + 1):
-                    with self.subTest(protocol=prot, type=t):
-                        pickled = pickle.dumps(x, prot)
-                        restored = pickle.loads(pickled)
-                        self.assertEqual(x, restored)
+    @skipUnless(PEP_560, "pickle support was added with pep_560")
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]:
```suggestion
    @skipUnless(PEP_560, "pickle support was added with PEP 560")
```

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 4]
```diff
@@ -1474,179 +1479,244 @@ def test_total(self):
...
+        LI = L[int]
+        with self.assertRaises(TypeError):
+            LI[None]
+
+    def test_annotated_in_other_types(self):
+        X = List[Annotated[T, 5]]
+        self.assertEqual(X[int], List[Annotated[int, 5]])
+
+
+@skipUnless(PEP_560, "python 3.7 required")
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]:
```suggestion
@skipUnless(PEP_560, "Python 3.7 required")
```

File: typing_extensions/src_py3/typing_extensions.py [hunk 5]
```diff
@@ -1653,6 +1661,11 @@ def __eq__(self, other):
         def __hash__(self):
             return hash((self.__origin__, self.__metadata__))
 
+        def __instancecheck__(self, obj):
+            raise TypeError("Annotated cannot be used with isinstance().")
+
+        def __subclasscheck__(self, cls):
+            raise TypeError("Annotated cannot be used with issubclass().")
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Why are these needed? To tweak the error message? I think this would be an error already with the inherited ones. Please add a comment.

File: typing_extensions/src_py3/typing_extensions.py [hunk 6]
```diff
@@ -1680,19 +1693,21 @@ class Annotated:
...
-            Optimized[int] == Annotated[int, runtime.Optimize]
+            Optimized = Annotated[T, runtime.Optimize()]
+            Optimized[int] == Annotated[int, runtime.Optimize()]
+
+            OptimizedList = Annotated[List[T], runtime.Optimize()]
+            OptimizedList[int] == Annotated[List[int], runtime.Optimize()]
 
-            OptimizedList = Annotated[List[T], runtime.Optimize]
-            OptimizedList[int] == Annotated[List[int], runtime.Optimize]
+        NOTE: the __type__ field is only available in python 3.7+.
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: I think this is now irrelevant, since we decide to not go with `__type__`.

File: typing_extensions/src_py3/typing_extensions.py [hunk 7]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
         if include_extras:
             return hint
         return {k: _strip_annotations(t) for k, t in hint.items()}
+
+elif HAVE_ANNOTATED:
+
+    # Prior to python 3.7 types did not have `copy_with`. A lot of the equality
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]:
```suggestion
    # Prior to Python 3.7 types did not have `copy_with`. A lot of the equality
```

File: typing_extensions/src_py3/typing_extensions.py [hunk 8]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+    # checks, argument expansion etc.. are done on the _subs_tree.
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]:
```suggestion
    # checks, argument expansion etc. are done on the _subs_tree.
```

File: typing_extensions/src_py3/typing_extensions.py [hunk 9]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+    #
+    # >> U = Union[Optional[float], int]
+    # >> U1 = Union[T, int]
+    # >> U2 = U1[Optional[float]]
+    # >> U == U2
+    # True
+    # >> U2.__args__, U.__args__
+    # (typing.Union[float, NoneType],) (<class 'float'>, <class 'NoneType'>, <class 'int'>)
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: I would remove these examples, they are not really helpful. Maybe instead mention which docstrings in `typing` to read?

File: typing_extensions/src_py3/typing_extensions.py [hunk 10]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+    #
+    # Here U and U2 have different shapes but have the same _subs_tree.
+    #
+    # As a result we can't provide a get_type_hints function that strips out
+    # annotations or a '__type__' field on the Annotated class.
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Again, this looks irrelevant, there is no `__type__`.

File: typing_extensions/src_py3/typing_extensions.py [hunk 11]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+                raise TypeError("Cannot subclass " + str(Annotated))
+            return super().__new__(cls, name, bases, namespace, **kwargs)
+
+        @property
+        def __metadata__(self):
+            return self._subs_tree()[2]
+
+        def _tree_repr(self, tree):
+            assert len(tree) == 3
+            # First argument is this class, second is __type__, 3rd is __metadata__
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Should this be `__type__` -> `__origin__`?

File: typing_extensions/src_py3/typing_extensions.py [hunk 12]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+                return Annotated
+            res = super()._subs_tree(tvars=tvars, args=args)
+            # Flatten nested Annotated
+            if isinstance(res[1], tuple) and res[1][0] is Annotated:
+                sub_tp = res[1][1]
+                sub_annot = res[1][2]
+                return (Annotated, sub_tp, sub_annot + res[2])
+            return res
+
+
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Please leave single empty line.

File: typing_extensions/src_py3/typing_extensions.py [hunk 13]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+        def _get_cons(self, msg):
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Add docstring for this method.

File: typing_extensions/src_py3/typing_extensions.py [hunk 14]
```diff
@@ -1755,3 +1772,165 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+                type_error = "a type variable."
+
+            if type_error:
+                raise TypeError(
+                    "{} {}: the annotated type is {}".format(
+                        msg,
+                        self,
+                        type_error
+                    )
+                )
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: Please don't use this style, just put it all on the same line with string, it will be more readable.

**Comment by ilevkivskyi** [comment]
> this diff is hard to read in this interface because it doesn't seem to offer a way to ignore whitespaces.

On the "Files changed tab" there is a little gear icon just above the diff, click on it, and check the box "Ignore whitespace".

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 15]
```diff
@@ -11,13 +11,21 @@
...
 except ImportError:
     pass
+try:
+    from typing_extensions import Annotated
+except ImportError:
+    pass
+try:
+    from typing_extensions import get_type_hints
+except ImportError:
+    from typing import get_type_hints
```
[the remark this answers is not in the document]
Comment (till-varoquaux) [reply typing_extensions/src_py3/test_typing_extensions.py]: get_type_hints only exists in `typing_extensions` for python3.7 and above (we can't trip out the annotations before that). I can re-alias it for the older versions if you prefer.

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 16]
```diff
@@ -1474,179 +1479,246 @@ def test_total(self):
...
-
-        def test_instantiate(self):
-            class C:
-                classvar = 4
+    def test_cannot_getattr_typevar(self):
+        with self.assertRaises(AttributeError):
+            Annotated[T, (5, 7)].x
 
-                def __init__(self, x):
-                    self.x = x
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]: Keep single empty line between methods.

File: typing_extensions/src_py3/test_typing_extensions.py [hunk 17]
```diff
@@ -1474,179 +1479,246 @@ def test_total(self):
...
-                {'other': MySet[T], 'return': MySet[T]}
-            )
+        self.assertEqual(L[int], Annotated[List[int], dec])
+        with self.assertRaises(TypeError):
+            L[int, int]
+
+        self.assertEqual(S[L[int]], Annotated[List[int], dec, dec2])
+
+
+
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/test_typing_extensions.py]: Why _three_ empty lines? I think one would be enough.

File: typing_extensions/src_py3/typing_extensions.py (hunk 5 above)
Comment (ilevkivskyi) [reply typing_extensions/src_py3/typing_extensions.py]: You marked this as resolved, but nothing has changed here, only in the other version (where it is actually more clear).

File: typing_extensions/src_py3/typing_extensions.py [hunk 18]
```diff
@@ -1755,3 +1770,170 @@ def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
...
+                cons = tree
+            type_error = None
+            if isinstance(cons, typing._ForwardRef):
+                type_error = "a forward reference."
+            elif isinstance(cons, typing.TypeVar):
+                type_error = "a type variable."
+
+            if type_error:
+                error = "{} {}: the annotated type is {}".format(msg, self, type_error)
+                raise exc_type(error)
```
Comment (ilevkivskyi) [inline typing_extensions/src_py3/typing_extensions.py]: TBH I think this would be cleaner if this method just returns `cons`, while the caller would raise an appropriate error and use another helper method to construct the error message.

More generally, why do we actually need this? If this method just returns a `TypeVar`, then a user will see exactly the same error as if one would try to instantiate it directly, etc.

File: typing_extensions/src_py3/test_typing_extensions.py (hunk 16 above)
Comment (ilevkivskyi) [reply typing_extensions/src_py3/test_typing_extensions.py]: GitHub screwed up this comment. To see it properly, apply the "Ignore whitespace" option.

File: typing_extensions/src_py3/typing_extensions.py (hunk 5 above)
Comment (till-varoquaux) [reply typing_extensions/src_py3/typing_extensions.py]: Oh, sorry. I thought I had added a comment: without this we get an error about "subscripted generics" which I thought was a bit confusing. Come to think of it though I've probably added too much code to control what error messages get thrown. Both this and get_cons can be simplified.
